### PR TITLE
SCMOD-8516: Disable TLS 1.1 and 1.2

### PR DESCRIPTION
**Pull request by gaana-vishwnath**
https://portal.digitalsafe.net/browse/SCMOD-8516

Continuation of #8.

#Cipher found for TLSv1.2 in https://wiki.mozilla.org/Security/Server_Side_TLS#Intermediate_compatibility_.28recommended.29
#Cipher found for TLSv1.2 in https://tools.ietf.org/html/rfc5246
#Cipher found for TLSv1.1 in https://tools.ietf.org/html/rfc4346#page-53

### CODE REVIEW

**Review by dermot-hardy — COMMENTED** [review]
You need to update the release notes. We need to get back into the practice of doing that. Apart from that this change looks OK. I see you have disabled further ciphers but have you confirmed that this is the complete list of ciphers that need to be disabled?

**Comment by gaana-vishwnath** [comment]
> You need to update the release notes. We need to get back into the practice of doing that. Apart from that this change looks OK. I see you have disabled further ciphers but have you confirmed that this is the complete list of ciphers that need to be disabled?

Updated the release notes. I used the  https://wiki.mozilla.org/Security/Server_Side_TLS#Modern_compatibility link to add the tlsv1.2 ciphers and also found few in https://tools.ietf.org/html/rfc5246#appendix-A.5. But I am not sure if that is the complete list.

**Comment by michael-bryson** [comment]
There is no update to the releases notes showing in here.

File: release-notes-2.1.0.md [hunk 1]
```diff
@@ -4,5 +4,6 @@
 ${version-number}
 
 #### New Features
+- **SCMOD-8516**: Extend the TLS/SSL cipher hardening of the Java base images to disallow tlsv1, tlsv1.1 and tlsv1.2
```
Comment (michael-bryson) [inline release-notes-2.1.0.md]: tls should be uppercase

Comment (gaana-vishwnath) [reply release-notes-2.1.0.md]: done

File: src/main/docker/disableWeakTlsAlgorithms.patch [hunk 2]
```diff
@@ -108,7 +110,90 @@
 +    TLS_DHE_RSA_WITH_CAMELLIA_256_CBC_SHA, \
 +    TLS_DHE_RSA_WITH_CAMELLIA_256_CBC_SHA256, \
 +    SSL_DHE_RSA_WITH_DES_CBC_SHA, \
-+    TLS_DHE_RSA_WITH_SEED_CBC_SHA
++    TLS_DHE_RSA_WITH_SEED_CBC_SHA, \
+#Cipher found for TLSv1.1 in https://tools.ietf.org/html/rfc4346#page-53
```
Comment (dermot-hardy) [inline src/main/docker/disableWeakTlsAlgorithms.patch]: Remove this line comment

File: src/main/docker/disableWeakTlsAlgorithms.patch [hunk 3]
```diff
@@ -108,7 +110,90 @@
...
++    TLS_KRB5_WITH_RC4_128_SHA, \
++    TLS_RSA_WITH_AES_128_CBC_SHA, \
++    TLS_RSA_WITH_AES_256_CBC_SHA, \
++    TLS_RSA_WITH_DES_CBC_SHA, \
++    TLS_RSA_WITH_IDEA_CBC_SHA, \
++    TLS_RSA_WITH_NULL_MD5, \
++    TLS_RSA_WITH_NULL_SHA, \
++    TLS_RSA_WITH_RC4_128_MD5, \
++    TLS_RSA_WITH_RC4_128_SHA, \
+#Cipher found for TLSv1.2 in https://wiki.mozilla.org/Security/Server_Side_TLS#Intermediate_compatibility_.28recommended.29
```
Comment (dermot-hardy) [inline src/main/docker/disableWeakTlsAlgorithms.patch]: Remove this line comment

File: src/main/docker/disableWeakTlsAlgorithms.patch [hunk 4]
```diff
@@ -108,7 +110,90 @@
...
++    ECDHE-ECDSA-AES128-GCM-SHA256, \
++    ECDHE-RSA-AES128-GCM-SHA256, \
++    ECDHE-ECDSA-AES256-GCM-SHA384, \
++    ECDHE-RSA-AES256-GCM-SHA384, \
++    ECDHE-ECDSA-CHACHA20-POLY1305, \
++    ECDHE-RSA-CHACHA20-POLY1305, \
++    DHE-RSA-AES128-GCM-SHA256, \
++    DHE-RSA-AES256-GCM-SHA384, \
+# Cipher found for TLSv1.2 in https://tools.ietf.org/html/rfc5246
```
Comment (dermot-hardy) [inline src/main/docker/disableWeakTlsAlgorithms.patch]: Remove this line comment

File: src/main/docker/disableWeakTlsAlgorithms.patch [hunk 5]
```diff
@@ -108,7 +110,90 @@
...
++    TLS_DH_DSS_WITH_AES_256_CBC_SHA256, \
++    TLS_DH_RSA_WITH_AES_256_CBC_SHA256, \
++    TLS_DHE_DSS_WITH_AES_256_CBC_SHA256, \
++    TLS_DHE_RSA_WITH_AES_256_CBC_SHA256, \
++    TLS_DH_anon_WITH_RC4_128_MD5, \
++    TLS_DH_anon_WITH_3DES_EDE_CBC_SHA, \
++    TLS_DH_anon_WITH_AES_128_CBC_SHA, \
++    TLS_DH_anon_WITH_AES_256_CBC_SHA, \
++    TLS_DH_anon_WITH_AES_128_CBC_SHA256, \
++    TLS_DH_anon_WITH_AES_256_CBC_SHA256
```
Comment (dermot-hardy) [inline src/main/docker/disableWeakTlsAlgorithms.patch]: There are a lot of ciphers being listed here.  Could a wildcard be safely used?

**Review by dermot-hardy — CHANGES_REQUESTED** [review]
Update release notes and version number to reflect that this is a breaking change.

**Review by dermot-hardy — APPROVED** [review]

**Comment by andyreidz** [comment]
It looks like Dermot wanted Rob to review this.

image

**Review by rob-duewer — COMMENTED** [review]
See comments left on https://github.com/CAFapi/opensuse-java8-images/pull/14.

But there are a couple differences with j11 vs. j8

You need to still permit the intermediate list of ciphers (so 3 for TLS 1.3, the other safe 8 for TLS 1.2 (assuming J11 supports chaha stream cipher, I haven't looked) as long as we have any images anchored to j8.  Only after we have moved every one of our ms's to java11, should we move to excluding **everything** but the 3 current safe "modern" TLS 1.3 ciphers.  Until then, enforce the moderate safety list.

Between https://docs.oracle.com/en/java/javase/11/security/oracle-providers.html#GUID-7093246A-31A3-4304-AC5F-5FB6400405E2 and https://docs.oracle.com/en/java/javase/11/docs/specs/security/standard-names.html#jsse-cipher-suite-names, you should be able to suss out what the approropate exclusion list (explicit and shortcut) should be for our j11 use.

**Pull request merged**